---
title: "Article (Data)"
author: "Alyssa Imbert and Etienne Thevenot (ProMetIS consortium)"
date: "`r doc_date()`"

vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: "ProMetIS.bib"
output:
  BiocStyle::html_document:
    toc: yes
    toc_depth: 4
    toc_float:
      collapsed: false
---

![](figures/prometis_logo.png)

```{r options, include=FALSE}
knitr::opts_chunk$set(fig.width = 8,
                      fig.height = 8,
                      fig.path = 'figures/temp/6_article_data/',
                      message = FALSE,
                      warning = FALSE)
```

# Loading the LAT and MX2 data

```{r loading}
mset.ls <- lapply(ProMetIS::genes.vc(),
                  function(gene.c) {
                    gene.mset <- phenomis::reading(file.path(ProMetIS::aggregated_dir.c(gene.c)),
                                                   report.c = "none")
                    gene.mset <- gene.mset[, ProMetIS::sets.vc()]
                  })
names(mset.ls) <- ProMetIS::genes.vc()
```

# Number of samples and features for each dataset

```{r dims, fig.width = 10}
data_dims.mn <- NULL
for (gene.c in ProMetIS::genes.vc()) {
  gene.mset <- mset.ls[[gene.c]]
  gene_dims.mn <- t(sapply(names(gene.mset),
                          function(set.c) {
                            eset <- gene.mset[[set.c]]
                            rev(dim(eset))
                          }))
  colnames(gene_dims.mn) <- paste0(gene.c, "_", colnames(gene_dims.mn))
  data_dims.mn <- cbind(data_dims.mn, gene_dims.mn)
}

data_dims.mn <- data_dims.mn[c("preclinical",
                               "proteomics_liver",
                               "metabolomics_liver_c18hyper_pos",
                               "metabolomics_liver_hilic_neg",
                               "proteomics_plasma",
                               "metabolomics_plasma_c18hyper_pos",
                               "metabolomics_plasma_hilic_neg",
                               "metabolomics_plasma_c18acqui_pos",
                               "metabolomics_plasma_c18acqui_neg"), ]

rownames(data_dims.mn) <- c("preclinical",
                            "liver_proteomics",
                            "liver_metabo_c18hyper_pos",
                            "liver_metabo_hilic_neg",
                            "plasma_proteomics",
                            "plasma_metabo_c18hyper_pos",
                            "plasma_metabo_hilic_neg",
                            "plasma_metabo_c18acqui_pos",
                            "plasma_metabo_c18acqui_neg")

p <- phenomis::gg_barplot(data_dims.mn,
                          bar_just.n = 0,
                          cex_bar.i = 5,
                          cex_axis.i = 14,
                          palette.vc = c(LAT_Samples = as.character(ProMetIS::palette.vc(light.l = TRUE)["LAT"]),
                                         LAT_Features = as.character(ProMetIS::palette.vc()["LAT"]),
                                         MX2_Samples = as.character(ProMetIS::palette.vc(light.l = TRUE)["MX2"]),
                                         MX2_Features = as.character(ProMetIS::palette.vc()["MX2"])),
                          figure.c = "none")
p <- p + ggplot2::scale_y_continuous(trans = "log10", limits = c(1, NA),
                                     expand = ggplot2::expansion(add = c(0, 1)))
print(p)

grDevices::png("figures/article_data/sets_dims.png",
               width = 740)

print(p)

grDevices::dev.off()
```

# Preclinical features

```{r preclinical}
preclinical.eset <- mset.ls[["LAT"]][["preclinical"]]

p <- phenomis::gg_pie(Biobase::fData(preclinical.eset), y.c = "category",
                 title.c = "Preclinical variables",
                 geom_text.ls = list(lab.i = 4, legend_text.i = 8, title.i = 18),
                 label.c = "value")

grDevices::png("figures/article_data/preclinical_categories.png")

print(p)

grDevices::dev.off()
```

# Technical validation

For each dataset, numerical and graphical quality controls are presented. 

## Preclinical (PHENOMIN ICS)

The objective is to check that the WT lines from the ProMetIS study have similar phenotyping values compared to all other WT lines generated by PHENOMIN ICS.

The ICS phenotyping database of WT lines is uploaded.

```{r}
wtdb_file.c <- "//10.0.238.33/Data/Phenostore/data/ProMetIS/cs1_phenomin/0_raw/technical_validation/preclinical/impc_5490_4sept2020.xlsx"

wtdb.tb <- suppressMessages(readxl::read_excel(wtdb_file.c,
                                               skip = 1))

wtdb.df <- as.data.frame(wtdb.tb,
                         stringsAsFactors = FALSE)

# excluding the first project (2 first rows) with NAs only

wtdb.df <- wtdb.df[-c(1, 2), ]

# naming rows

rownames(wtdb.df) <- paste0(wtdb.df[, "project"], "_", wtdb.df[, "mice"], "_", wtdb.df[, "sex"])
```

### Reference range

```{r}
message("Total nb of WT mice: ", nrow(wtdb.df))
message("Nb of females and males: ")
print(table(wtdb.df[, "sex"]))
```

```{r}
num.vl <- sapply(wtdb.df, data.class) == "numeric"

message("Number of numerical variables: ", sum(num.vl))

wtdb_num.df <- wtdb.df[, num.vl]
wtdb_num.ls <- split(wtdb_num.df, wtdb.df[, "sex"])

wtdb_range.ls <- lapply(wtdb_num.ls,
                        function(wtdb.mn) {
                          mean.vn <- apply(wtdb.mn, 2, function(x) mean(x, na.rm = TRUE))
                          sd.vn <- apply(wtdb.mn, 2, function(x) sd(x, na.rm = TRUE))
                          rbind(min = mean.vn - 2 * sd.vn,
                                max = mean.vn + 2 * sd.vn)
                        })

project.vi <- which(wtdb.df[, "project"] == "5490bp")
wtpm_mean.mn <- apply(wtdb_num.df[project.vi, ], 2,
                      function(y) tapply(y,
                                         wtdb.df[project.vi, "sex"],
                                         function(x) mean(x, na.rm = TRUE)))

wtdb_inrange.ls <- vector(mode = "list", length = length(wtdb_num.ls))
names(wtdb_inrange.ls) <- names(wtdb_num.ls)

for (sex.c in names(wtdb_inrange.ls)) {
  
  wtdb.mn <- wtdb_num.ls[[sex.c]]
  
  wtpm_inrange.vl <- logical(ncol(wtpm_mean.mn))
  names(wtpm_inrange.vl) <- colnames(wtpm_mean.mn)
  for (j in 1:ncol(wtpm_mean.mn))
    wtpm_inrange.vl[j] <- wtdb_range.ls[[sex.c]]["min", j] <= wtpm_mean.mn[sex.c, j] &
      wtpm_mean.mn[sex.c, j] <= wtdb_range.ls[[sex.c]]["max", j]
  
  wtdb_inrange.ls[[sex.c]] <- wtpm_inrange.vl
}

for (sex.c in names(wtdb_inrange.ls)) {
  
  message("Sex: ", ifelse(sex.c == "F", "females", "males"))
  wtpm_inrange.vl <- wtdb_inrange.ls[[sex.c]]
  
  wtpm_outlier.vi <- which(!wtpm_inrange.vl)
  message("Oulier variables: ")
  print(wtpm_outlier.vi)
  message("Non outlier variables in males: ",
          signif(sum(wtpm_inrange.vl, na.rm = TRUE) /
                   length(wtpm_inrange.vl[!is.na(wtpm_inrange.vl)]) * 100, 3),
          "%")
  
}
```


## Proteomics


```{r proteomics_files}
# liver
prot_liver_file.c <- "//10.0.238.33/Data/Phenostore/data/ProMetIS/cs1_phenomin/0_raw/technical_validation/proteomics_liver/DonnÃ©es pour figures Strasbourg QC.xlsx"

# plasma
prot_plasma_file.c <- "//10.0.238.33/Data/Phenostore/data/ProMetIS/cs1_phenomin/0_raw/technical_validation/proteomics_liver/figureQC_proteomique.xlsx"
```

### iRT

```{r proteomics_irt_loading}
# liver
irt_liver.tb <- suppressMessages(readxl::read_excel(prot_liver_file.c,
                                                     sheet = 1))

irt_liver.df <- as.data.frame(irt_liver.tb,
                               stringsAsFactors = FALSE)

irt_liver.mn <- as.matrix(irt_liver.df[-1, -1])
mode(irt_liver.mn) <- "numeric"
rownames(irt_liver.mn) <- irt_liver.df[, 1][-1]

pept_liver.vc <- rownames(irt_liver.mn)
rownames(irt_liver.mn) <- paste0("P", 1:nrow(irt_liver.mn))

# plasma

irt_plasma.tb <- suppressMessages(readxl::read_excel(prot_plasma_file.c,
                                                      sheet = 2))

irt_plasma.df <- as.data.frame(irt_plasma.tb,
                                stringsAsFactors = FALSE)

irt_plasma.mn <- as.matrix(irt_plasma.df[3:13, 21:65])
mode(irt_plasma.mn) <- "numeric"
rownames(irt_plasma.mn) <- irt_plasma.df[, 2][3:13]
pept_plasma.vc <- rownames(irt_plasma.mn)
stopifnot(identical(pept_liver.vc, pept_plasma.vc))
rownames(irt_plasma.mn) <- rownames(irt_liver.mn)
```

### CV

```{r prot_cv_load}
# liver
cv_liver.tb <- suppressMessages(readxl::read_excel(prot_liver_file.c,
                                                     sheet = 2,
                                                     range = "A1:E2469"))

cv_liver.df <- as.data.frame(cv_liver.tb,
                             stringsAsFactors = FALSE)

cv_liver.df <- cbind.data.frame(type = factor(rep(c("Pool", "WT", "Lat", "Mx2"),
                                                  each = nrow(cv_liver.df)),
                                              levels = c("Pool", "WT", "Lat", "Mx2")),
                                value = c(cv_liver.df[, 2], cv_liver.df[, 3],
                                          cv_liver.df[, 4], cv_liver.df[, 5]))

# plasma
cv_plasma.tb <- suppressMessages(readxl::read_excel(prot_plasma_file.c,
                                                      sheet = 4,
                                                      range = "A1:BT622"))

cv_plasma.df <- as.data.frame(cv_plasma.tb,
                                stringsAsFactors = FALSE)

cv_plasma.df <- cbind.data.frame(type = factor(rep(c("WT", "Mx2", "Lat", "Pool"),
                                                   each = nrow(cv_plasma.df)),
                                               levels = c("Pool", "WT", "Lat", "Mx2")),
                                 value = c(cv_plasma.df[, 69], cv_plasma.df[, 70],
                                           cv_plasma.df[, 71], cv_plasma.df[, 72]))
```

### Plotting

```{r}
irt_plot <- function(data.mn,
                     tissue.c = c("liver", "plasma")[1]) {
  
  data.df <- as.data.frame(t(data.mn))
  data.df[, "id"] <- 1:nrow(data.df)

  
  # reshaping in the long format
  plot_data.df <- reshape2::melt(data.df, id.var = "id")
  
  run.vc <- sapply(rownames(data.df),
                   function(x)
                     unlist(strsplit(unlist(strsplit(x, split = ".", fixed = TRUE))[1],
                                     split = "_"))[1],
                   USE.NAMES = FALSE)

  colnames(plot_data.df) <- gsub("variable", "peptides", colnames(plot_data.df))
  
  # plotting
  p <- ggplot2::ggplot(plot_data.df,
                       ggplot2::aes(x = id, y = value,
                                    group = peptides,
                                    colour = peptides)) +
    ggplot2::geom_point() +
    ggplot2::geom_line(ggplot2::aes(lty = peptides)) +
    ggplot2::scale_color_manual(values = RColorBrewer::brewer.pal(12, "Paired")[-11]) +
    # ggplot2::scale_color_brewer(palette = "Paired") +
    ggplot2::theme_light() +
    ggplot2::labs(title = tissue.c,
                  x = "LC-MS run",
                  y = "Retention time") +
    ggplot2::theme(axis.text = ggplot2::element_text(size = 11, face = "bold"),
                   axis.title = ggplot2::element_text(size = 11, face = "bold"),
                   axis.text.x = ggplot2::element_text(angle = 90,
                                                       size = 9,
                                                       hjust = 1,
                                                       vjust = 0.5),
                   legend.title = ggplot2::element_text(face = "bold"),
                   legend.text = ggplot2::element_text(face = "bold")) +
    ggplot2::scale_x_continuous(breaks = 1:length(run.vc),
                                labels = run.vc,
                                name = "LC-MS run")
                             
  return(invisible(p))
  
}




irt_liver.gg <- irt_plot(irt_liver.mn)
irt_plasma.gg <- irt_plot(irt_plasma.mn, "plasma")
cv_liver.gg <- .cv_ggplot(cv_liver.df)
cv_plasma.gg <- .cv_ggplot(cv_plasma.df, "plasma")
```

```{r}
prot_qc_plot <- gridExtra::grid.arrange(irt_liver.gg,
                                        cv_liver.gg,
                                        irt_plasma.gg,
                                        cv_plasma.gg,
                                        nrow = 2, ncol = 2,
                                        widths = c(4, 2),
                                        heights = c(3, 3))
ggplot2::ggsave("figures/article_data/proteo_irt_cv.pdf", prot_qc_plot, width = 12, height = 7)
```

## Metabolomics

Loading the processed datasets:

```{r metabosets_loading}
metabo.mset <- phenomis::reading(file.path(ProMetIS::processed_dir.c()),
                                 subsets.vc = ProMetIS::metabo_sets.vc(),
                                 report.c = "none")
```

Post-processing in the 'technical validation' mode (keeping standards, keeping pools, omitting the pool_CV <= 30% and pool_CV_over_sample_CV <= 1 filters)

```{r metabosets_postprocessing_technical_validation}
metabo.mset <- ProMetIS:::.metabo_postprocessing(metabo.mset = metabo.mset,
                                                 drift_correct.c = "prometis",
                                                 .technical_validation.l = TRUE)
```

### Quality metrics

```{r metabosets_quality_metrics}
quality_metrics.ls <- ProMetIS:::.metabo_quality_metrics(metabo.mset)
```

```{r}
knitr::kable(round(quality_metrics.ls[["quality.mn"]]))
```

```{r inj_order_radarplot}
quality.df <- as.data.frame(t(as.matrix(quality_metrics.ls[["quality.mn"]][c("correl_inj_test",
                                                                             "correl_inj_pca",
                                                                             "pool_spread_pca",
                                                                             "poolCV_0.3",
                                                                             "ICCpool"), ])))
colnames(quality.df) <- c("spearman", "pca", "spread", "cv", "icc")
quality.df <- cbind.data.frame(group = rownames(quality.df),
                               quality.df)
for (metric.c in c("spearman", "pca", "spread"))
  quality.df[, metric.c] <- 100 - quality.df[, metric.c]

quality.ggradar <- ggradar::ggradar(quality.df,
                                    base.size = 12,
                                    values.radar = c("0", "50", "100"),
                                    grid.min = 0, grid.mid = 50, grid.max = 100,
                                    # Polygones
                                    group.line.width = 2, 
                                    group.point.size = 3,
                                    group.colours = RColorBrewer::brewer.pal(9, "Set1")[c(1:5, 7)],
                                    # Background
                                    background.circle.colour = "white",
                                    gridline.mid.colour = "grey",
                                    # Legend
                                    plot.title = "Quality metrics",
                                    legend.text.size = 10,
                                    legend.position = "bottom")


ggplot2::ggsave("figures/article_data/metabo_injOrder_radar.pdf",
                quality.ggradar,
                height = 10.5, width = 7)

quality.mn <- quality.df[, -1]
print(round(quality.mn))
print(round(apply(quality.mn, 1, function(x) c(min = min(x), mean = mean(x), max = max(x)))))
```

### Score plots

```{r metabosets_quality_scoreplots}
scoreplot.ls <- vector(mode = "list", length = length(ProMetIS::metabo_sets.vc()))
names(scoreplot.ls) <- ProMetIS::metabo_sets.vc()

for (set.c in names(metabo.mset)) {
  
  eset <- metabo.mset[[set.c]]
  
  eset <- eset[, Biobase::pData(eset)[, "sampleType"] %in% c("pool", "sample")]
  
  pdata.df <- Biobase::pData(eset)
  pdata.df[, "order"] <- pdata.df[, "injectionOrder"]
  if (grepl("acqui", set.c))
    pdata.df[, "order"] <- pdata.df[, "order"] - min(pdata.df[, "order"]) + 1
  pdata.df[, "label"] <- ifelse(pdata.df[, "sampleType"] == "pool", "QC", "s")
  Biobase::pData(eset) <- pdata.df
  
  eset.pca <- ropls::opls(eset, predI = 2, fig.pdfC = "none", info.txtC = "none")
  
  scoreplot.ls[[set.c]] <- ProMetIS:::score_plotly(eset.pca,
                                                   label.c = "label",
                                                   color.c = "order",
                                                   title.c = gsub("metabolomics_", "", set.c),
                                                   figure.c = "interactive",
                                                   size.ls = list(axis_lab.i = 12,
                                                                  axis_text.i = 10,
                                                                  point.i = 2,
                                                                  label.i = 4,
                                                                  title.i = 15,
                                                                  legend_title.i = 10,
                                                                  legend_text.i = 10),
                                                   plot.l = FALSE)
  
}

scoreplots.p <- gridExtra::grid.arrange(grobs = scoreplot.ls[c("metabolomics_liver_c18hypersil_pos",
                                                               "metabolomics_plasma_c18hypersil_pos",
                                                               "metabolomics_plasma_c18acquity_pos",
                                                               "metabolomics_liver_hilic_neg",
                                                               "metabolomics_plasma_hilic_neg",
                                                               "metabolomics_plasma_c18acquity_neg")],
                                        nrow = 2)

ggplot2::ggsave(paste0("figures/article_data/metabo_qc.pdf"), scoreplots.p,
                height = 7, width = 10.5)
```

### Drift

```{r metabo_drifts}
pdf("figures/article_data/metabo_drifts.pdf",
    height = 7, width = 10.5)

layout(mat = matrix(1:6, nrow = 2),
       heights = c(3.5, 3.5),
       widths = c(3.5, 3.5, 3.5))

for (set.c in names(metabo.mset)) {
  eset <- metabo.mset[[set.c]]
  ProMetIS:::.metabo_drift_plot(eset, title.c = gsub("metabolomics_", "", set.c))
}

dev.off()
```

### CV

```{r metabo_cv}
cv_ggplot.ls <- vector(mode = "list", length = length(names(metabo.mset)))
names(cv_ggplot.ls) <- names(metabo.mset)

for (set.c in names(metabo.mset)) {
  
  # set.c <- names(metabo.mset)[2]
  eset <- metabo.mset[[set.c]]
  eset <- eset[, Biobase::pData(eset)[, "sampleType"] %in% c("pool", "sample")]
  
  genotype.vc <- tolower(substr(Biobase::pData(eset)[,
                                                     ifelse(grepl("(hyper|hilic)", set.c),
                                                            "KO", "Type")], 1, 1))
  
  type.vc <- vapply(seq_len(Biobase::dims(eset)["Samples", 1]),
                    function(sample.i) {
                      if (Biobase::pData(eset)[sample.i, "sampleType"] == "pool") {
                        return("Pool")
                      } else if (genotype.vc[sample.i] == "l") {
                        return("Lat")
                      } else if (genotype.vc[sample.i] == "m") {
                        return("Mx2")
                      } else {
                        return("WT")
                      }
                    }, FUN.VALUE = character(1))
  cv.mn <- 100 * as.matrix(t(apply(Biobase::exprs(eset), 1,
                                   function(feat.vn) {
                                     tapply(feat.vn, type.vc,
                                            function(x) sd(x, na.rm = TRUE) / mean(x, na.rm = TRUE))
                                   })))
  
  cv.df <- cbind.data.frame(type = factor(rep(c("Pool", "WT", "Lat", "Mx2"),
                                              each = nrow(cv.mn)),
                                          levels = c("Pool", "WT", "Lat", "Mx2")),
                            value = c(cv.mn[, "Pool"], cv.mn[, "WT"],
                                      cv.mn[, "Lat"], cv.mn[, "Mx2"]))
  
  cv_ggplot.ls[[set.c]] <- ProMetIS:::.cv_ggplot(cv.df, title.c = gsub("metabolomics_", "", set.c))
  
}

metabo_cv_plot <- gridExtra::marrangeGrob(cv_ggplot.ls,
                                          nrow = 2, ncol = 3)

ggplot2::ggsave(paste0("figures/article_data/metabo_cv.pdf"), metabo_cv_plot,
                height = 7, width = 10.5)
```


```{r metabo_quality_cv_percent}
pdf("figures/article_data/metabo_quality_cv_percent.pdf",
    height = 7, width = 10.5)

ProMetIS:::.metabo_quality_plot(quality_metrics.ls = quality_metrics.ls, "cv_percent")

dev.off()
```

### ICC

```{r metabo_quality_icc_intens}
pdf("figures/article_data/metabo_quality_icc_intens.pdf",
    height = 7, width = 10.5)
    
ProMetIS:::.metabo_quality_plot(quality_metrics.ls = quality_metrics.ls, "icc_intens")

dev.off()
```

### Standards


```{r metabo_standards_matching}
for (set.c in grep("(c18hyper|hilic)", names(metabo.mset), value = TRUE)) {
  
  # set.c <- grep("(c18hyper|hilic)", names(metabo.mset), value = TRUE)[2]
  
  eset <- metabo.mset[[set.c]]
  eset <- eset[Biobase::fData(eset)[, "standard"] != "",
               Biobase::pData(eset)[, "sampleType"] %in% c("pool", "sample")]
  
  #ggplot needs a dataframe
  exprs.mn <- Biobase::exprs(eset)
  
  tlog2exprs.df <- as.data.frame(t(log2(1 + exprs.mn)))
  colnames(tlog2exprs.df) <- Biobase::fData(eset)[, "standard"]
  
  stand_unselect.vc <- switch(set.c,
                              metabolomics_liver_c18hypersil_pos = {
                                ""
                              },
                              metabolomics_liver_hilic_neg = {
                                "Prednisone [ExS]"
                              },
                              metabolomics_plasma_c18hypersil_pos = {
                                c("2-Aminoanthracene [ExS]",
                                  "Alanine 13C [ExS]",
                                  "AMPA")
                              },
                              metabolomics_plasma_hilic_neg = {
                                c("Amiloride [ExS]",
                                  "Ethylmalonic acid [ExS]")
                              })
  tlog2exprs.df <- tlog2exprs.df[, setdiff(colnames(tlog2exprs.df), stand_unselect.vc), drop = FALSE]
  
  #id variable for position in matrix 
  tlog2exprs.df$id <- 1:nrow(tlog2exprs.df) 
  #reshape to long format
  ggplot.df <- reshape2::melt(tlog2exprs.df, id.var = "id")
  
  ggplot.df[, "sample"] <- factor(rep(gsub("pool1", "pool",
                                           Biobase::pData(eset)[, "sampleType"]),
                                      ncol(tlog2exprs.df) - 1),
                                  levels = c("pool",
                                             "sample"))
  
  ggplot.df[, "compound_standard"] <- ggplot.df[, "variable"]
  ggplot.df[, "compound"] <- vapply(as.character(ggplot.df[, "compound_standard"]),
                                           function(comp.c) {
                                             bracket.i <- which(unlist(strsplit(comp.c, split = "")) == "[")
                                             substr(comp.c, 1, bracket.i - 2)
                                           }, character(1))
  ggplot.df[, "standard"] <- vapply(as.character(ggplot.df[, "compound_standard"]),
                                    function(comp.c) {
                                      bracket.i <- which(unlist(strsplit(comp.c, split = "")) == "[")
                                      substr(comp.c, bracket.i, nchar(comp.c))
                                    }, character(1))
  ggplot.df[, "injectionOrder"] <- ggplot.df[, "id"]
  ggplot.df[, "intensity"] <- ggplot.df[, "value"]
  
  #plot
  stand.ggplot <- ggplot2::ggplot(ggplot.df,
                                  ggplot2::aes(x = injectionOrder,
                                               y = intensity,
                                               group = compound,
                                               colour = compound)) +
    ggplot2::geom_point(ggplot2::aes(shape = sample), size = 3) +
    ggplot2::geom_line(ggplot2::aes(lty = standard), size = 2) +
    ggplot2::labs(title = gsub("metabolomics_", "", set.c),
                  x = "Injection order", y = "Intensity") +
    ggplot2::scale_shape_manual(values = c(15, 16)) +
    ggplot2::scale_color_brewer(palette = "Dark2") +
    ggplot2::coord_cartesian(ylim = c(0, 30)) +
    ggplot2::theme_light()
  
  ggplot2::ggsave(paste0("figures/article_data/metabo_standards_", gsub("metabolomics_", "", set.c), ".pdf"),
                  stand.ggplot, width = 12)
  
}
```
